File: app/src/main/java/com/liarstudio/mvideosmsfilter/AndroidParser.java
package com.liarstudio.mvideosmsfilter;

import android.content.ContentResolver;
import android.content.Context;
import android.database.Cursor;
import android.net.Uri;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class AndroidParser extends Parser {

    final Uri SMS_INBOX = Uri.parse("content://sms/inbox");
    final static String SMS_NAME = "M.Video";

    private Context context;

    public AndroidParser(Context context) {
        this.context = context;
    }

/*    @Override
    List<String> parseTemplate(String code) {

        Pattern pattern = Pattern.compile(code);

        List<String> coupons = new ArrayList<>();

        String[] columnNames = {"_id", "body", "address", "date"};
        ContentResolver cr = context.getContentResolver();
        Cursor cursor = cr.query(SMS_INBOX, columnNames, "address='" + SMS_NAME + "'", null, "date asc");//, null, null);


        while (cursor.moveToNext()) {
            String current = cursor.getString(cursor.getColumnIndexOrThrow("body"));
            Matcher matcherCode = pattern.matcher(current);
            if (matcherCode.find())
                coupons.add(matcherCode.group() + "\r\n");
        }

        return coupons;
    }

*/

    @Override
    List<String> parseTemplate(RegexPattern... patterns) {

        List<String> coupons = new ArrayList<>();

        String[] columnNames = {"_id", "body", "address", "date"};
        ContentResolver cr = context.getContentResolver();
        Cursor cursor = cr.query(SMS_INBOX, columnNames, "address='" + SMS_NAME + "'", null, "date asc");//, null, null);

        while (cursor.moveToNext()) {
            String current = cursor.getString(cursor.getColumnIndexOrThrow("body"));
            String coupon = null;
            for (RegexPattern pattern : patterns) {
                coupon = parseTemplateCoupon(current, pattern);
                if (coupon != null)
                    coupons.add(coupon);
            }
        }
        return coupons;

    }
    String parseTemplateCoupon(String current, RegexPattern pattern)  {

        String[] extractedData = pattern.extract();


        Pattern patternCode = Pattern.compile(extractedData[0]);
        Matcher matcherCode = patternCode.matcher(current);

        if (pattern.getTask() == Task.SMS || pattern.getTask() == Task.Philips) {
            Pattern patternAmount = Pattern.compile(extractedData[1]);
            Pattern patternSum = Pattern.compile(extractedData[2]);
            Pattern patternCond = Pattern.compile(extractedData[3]);

            Matcher matcherAmount = patternAmount.matcher(current);
            Matcher matcherSum = patternSum.matcher(current);
            Matcher matcherCond = patternCond.matcher(current);

            if (matcherSum.find() && matcherCode.find() && matcherAmount.find() && matcherCond.find())
                return calculatePrice(matcherAmount.group()) + "/" +
                        calculatePrice(matcherSum.group()) + ": " +
                        matcherCode.group() + "\r\n";

        } else {
            if (matcherCode.find())
                return  matcherCode.group() + "\r\n";
        }
        return null;
    }
}

File: app/src/main/java/com/liarstudio/mvideosmsfilter/Parsers/AndroidParser.kt
package com.liarstudio.mvideosmsfilter.Parsers

import android.content.Context
import android.net.Uri

import java.util.ArrayList
import java.util.Calendar
import java.util.regex.Pattern

class AndroidParser(private val context: Context) : Parser() {

    internal val SMS_INBOX = Uri.parse("content://sms/inbox")


     override fun parseTemplate(date: Calendar, vararg patterns: RegexPattern): MutableList<String> {

        val coupons = ArrayList<String>()

        val cr = context.contentResolver

        val columnNames = arrayOf("_id", "body", "address", "date")
        val selection: String
        val selectionArgs: Array<String>
        val sortOrder = "date asc"
        if (date == null) {
            selection = "address=?"
            selectionArgs = arrayOf(SMS_NAME)
        } else {
            selection = "address=? AND date>?"
            selectionArgs = arrayOf(SMS_NAME, java.lang.Long.toString(date.timeInMillis))
        }

        val cursor = cr.query(SMS_INBOX, columnNames, selection, selectionArgs, sortOrder)//, null, null);

        while (cursor!!.moveToNext()) {
            val current = cursor.getString(cursor.getColumnIndexOrThrow("body"))
            var coupon: String?
            for (pattern in patterns) {
                coupon = parseTemplateCoupon(current, pattern)
                if (coupon != null)
                    coupons.add(coupon)
            }
        }
        return coupons

    }

    internal fun parseTemplateCoupon(current: String, pattern: RegexPattern): String? {

        val extractedData = pattern.extract()


        val patternCode = Pattern.compile(extractedData!![0])
        val matcherCode = patternCode.matcher(current)

        if (pattern.task === Task.SMS || pattern.task === Task.Philips) {
            val patternAmount = Pattern.compile(extractedData[1])
            val patternSum = Pattern.compile(extractedData[2])
            val patternCond = Pattern.compile(extractedData[3])

            val matcherAmount = patternAmount.matcher(current)
            val matcherSum = patternSum.matcher(current)
            val matcherCond = patternCond.matcher(current)

            if (matcherSum.find() && matcherCode.find() && matcherAmount.find() && matcherCond.find())
                return calculatePrice(matcherAmount.group()).toString() + "/" +
                        calculatePrice(matcherSum.group()) + ": " +
                        matcherCode.group() + "\r\n"

        } else {
            if (matcherCode.find())
                return matcherCode.group() + "\r\n"
        }
        return null
    }

    companion object {
        internal val SMS_NAME = "M.Video"
    }
}